From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thomas Taube <Thomastaube@gmx.de>
Date: Mon, 12 Oct 2026 16:46:17 +0200
Subject: [PATCH] giochannel: Fix memcmp() off the end of the buffer with long
 terminators

Fixes CVE-2026-58013. Backport of upstream commits 7c6b11df2 (line terminator
stays nul-terminated; prerequisite) and 6a2583dec (glib 2.88.1, glib#3925) to
2.82.5.

CVE: CVE-2026-58013
---
 glib/giochannel.c | 31 +++++++++++++++++++------------
 1 file changed, 19 insertions(+), 12 deletions(-)

diff --git a/glib/giochannel.c b/glib/giochannel.c
index b44fff3..640f698 100644
--- a/glib/giochannel.c
+++ b/glib/giochannel.c
@@ -877,26 +877,31 @@ g_io_channel_set_line_term (GIOChannel	*channel,
                             const gchar	*line_term,
 			    gint         length)
 {
-  guint length_unsigned;
-
   g_return_if_fail (channel != NULL);
   g_return_if_fail (line_term == NULL || length != 0); /* Disallow "" */
 
+  g_free (channel->line_term);
+
   if (line_term == NULL)
-    length_unsigned = 0;
+    {
+      channel->line_term = NULL;
+      channel->line_term_len = 0;
+    }
   else if (length >= 0)
-    length_unsigned = (guint) length;
+    {
+      /* We store the value nul-terminated even if the input is not */
+      channel->line_term = g_malloc0 (length + 1);
+      memcpy (channel->line_term, line_term, length);
+      channel->line_term_len = (guint) length;
+    }
   else
     {
-      /* FIXME: We’re constrained by line_term_len being a guint here */
+      /* We’re constrained by line_term_len being a guint here */
       gsize length_size = strlen (line_term);
       g_return_if_fail (length_size <= G_MAXUINT);
-      length_unsigned = (guint) length_size;
+      channel->line_term = g_strdup (line_term);
+      channel->line_term_len = (guint) length_size;
     }
-
-  g_free (channel->line_term);
-  channel->line_term = line_term ? g_memdup2 (line_term, length_unsigned) : NULL;
-  channel->line_term_len = length_unsigned;
 }
 
 /**
@@ -906,7 +911,8 @@ g_io_channel_set_line_term (GIOChannel	*channel,
  *
  * This returns the string that #GIOChannel uses to determine
  * where in the file a line break occurs. A value of %NULL
- * indicates autodetection.
+ * indicates autodetection. Since 2.84, the return value is always
+ * nul-terminated.
  *
  * Returns: The line termination string. This value
  *   is owned by GLib and must not be freed.
@@ -1822,7 +1828,8 @@ read_again:
         {
           if (channel->line_term)
             {
-              if (memcmp (channel->line_term, nextchar, line_term_len) == 0)
+              if ((size_t) (lastchar - nextchar) >= line_term_len &&
+                  memcmp (channel->line_term, nextchar, line_term_len) == 0)
                 {
                   line_length = nextchar - use_buf->str;
                   got_term_len = line_term_len;
